// infamous/html/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.useDefaultNames = exports.WebComponent = exports.HTMLScene = exports.HTMLPushPaneLayout = exports.HTMLNode = exports.DeclarativeBase = undefined;

var _DeclarativeBase = require('./DeclarativeBase');

var _DeclarativeBase2 = _interopRequireDefault(_DeclarativeBase);

var _HTMLNode = require('./HTMLNode');

var _HTMLNode2 = _interopRequireDefault(_HTMLNode);

var _HTMLPushPaneLayout = require('./HTMLPushPaneLayout');

var _HTMLPushPaneLayout2 = _interopRequireDefault(_HTMLPushPaneLayout);

var _HTMLScene = require('./HTMLScene');

var _HTMLScene2 = _interopRequireDefault(_HTMLScene);

var _PushPaneLayout = require('../components/PushPaneLayout');

var _PushPaneLayout2 = _interopRequireDefault(_PushPaneLayout);

var _WebComponent = require('./WebComponent');

var _WebComponent2 = _interopRequireDefault(_WebComponent);

var _Scene = require('../core/Scene');

var _Scene2 = _interopRequireDefault(_Scene);

var _Node = require('../core/Node');

var _Node2 = _interopRequireDefault(_Node);

var _Mesh = require('../core/Mesh');

var _Mesh2 = _interopRequireDefault(_Mesh);

var _Box = require('../core/Box');

var _Box2 = _interopRequireDefault(_Box);

var _Sphere = require('../core/Sphere');

var _Sphere2 = _interopRequireDefault(_Sphere);

var _Plane = require('../core/Plane');

var _Plane2 = _interopRequireDefault(_Plane);

var _PointLight = require('../core/PointLight');

var _PointLight2 = _interopRequireDefault(_PointLight);

var _DOMNode = require('../core/DOMNode');

var _DOMNode2 = _interopRequireDefault(_DOMNode);

var _DOMPlane = require('../core/DOMPlane');

var _DOMPlane2 = _interopRequireDefault(_DOMPlane);

var _AmbientLight = require('../core/AmbientLight');

var _AmbientLight2 = _interopRequireDefault(_AmbientLight);

var _Camera = require('../core/Camera');

var _Camera2 = _interopRequireDefault(_Camera);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function useDefaultNames() {
    // TODO replace with a loop
    if (!customElements.get(_Scene2.default.defaultElementName)) {
        _Scene2.default.define();
    }
    if (!customElements.get(_Node2.default.defaultElementName)) {
        _Node2.default.define();
    }
    if (!customElements.get(_Mesh2.default.defaultElementName)) {
        _Mesh2.default.define();
    }
    if (!customElements.get(_Box2.default.defaultElementName)) {
        _Box2.default.define();
    }
    if (!customElements.get(_Sphere2.default.defaultElementName)) {
        _Sphere2.default.define();
    }
    if (!customElements.get(_Plane2.default.defaultElementName)) {
        _Plane2.default.define();
    }
    if (!customElements.get(_PointLight2.default.defaultElementName)) {
        _PointLight2.default.define();
    }
    if (!customElements.get(_DOMNode2.default.defaultElementName)) {
        _DOMNode2.default.define();
    }
    if (!customElements.get(_DOMPlane2.default.defaultElementName)) {
        _DOMPlane2.default.define();
    }
    if (!customElements.get(_AmbientLight2.default.defaultElementName)) {
        _AmbientLight2.default.define();
    }
    if (!customElements.get(_Camera2.default.defaultElementName)) {
        _Camera2.default.define();
    }
    //PushPaneLayout.define()
}

exports.DeclarativeBase = _DeclarativeBase2.default;
exports.HTMLNode = _HTMLNode2.default;
exports.HTMLPushPaneLayout = _HTMLPushPaneLayout2.default;
exports.HTMLScene = _HTMLScene2.default;
exports.WebComponent = _WebComponent2.default;
exports.useDefaultNames = useDefaultNames;
