// infamous/html/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.useDefaultNames = exports.WebComponent = exports.HTMLScene = exports.HTMLPushPaneLayout = exports.HTMLNode = exports.DeclarativeBase = undefined;

var _DeclarativeBase = require('./DeclarativeBase');

var _DeclarativeBase2 = _interopRequireDefault(_DeclarativeBase);

var _HTMLNode = require('./HTMLNode');

var _HTMLNode2 = _interopRequireDefault(_HTMLNode);

var _HTMLPushPaneLayout = require('./HTMLPushPaneLayout');

var _HTMLPushPaneLayout2 = _interopRequireDefault(_HTMLPushPaneLayout);

var _HTMLScene = require('./HTMLScene');

var _HTMLScene2 = _interopRequireDefault(_HTMLScene);

var _WebComponent = require('./WebComponent');

var _WebComponent2 = _interopRequireDefault(_WebComponent);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

//import 'document-register-element'
function useDefaultNames() {
    _HTMLNode2.default.define();
    _HTMLScene2.default.define();
    _HTMLPushPaneLayout2.default.define();
}

exports.DeclarativeBase = _DeclarativeBase2.default;
exports.HTMLNode = _HTMLNode2.default;
exports.HTMLPushPaneLayout = _HTMLPushPaneLayout2.default;
exports.HTMLScene = _HTMLScene2.default;
exports.WebComponent = _WebComponent2.default;
exports.useDefaultNames = useDefaultNames;
