// infamous/motor/Curves.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _tween = require('tween.js');

var _tween2 = _interopRequireDefault(_tween);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {

    // Linear
    Linear: _tween2.default.Easing.Linear.None,

    // Exponential
    ExponentialIn: _tween2.default.Easing.Exponential.In,
    ExponentialOut: _tween2.default.Easing.Exponential.InOut,
    ExponentialInOut: _tween2.default.Easing.Exponential.InOut,

    // Quadratic
    QuadraticIn: _tween2.default.Easing.Quadratic.In,
    QuadraticOut: _tween2.default.Easing.Quadratic.InOut,
    QuadraticInOut: _tween2.default.Easing.Quadratic.InOut,

    // Cubic
    CubicIn: _tween2.default.Easing.Cubic.In,
    CubicOut: _tween2.default.Easing.Cubic.InOut,
    CubicInOut: _tween2.default.Easing.Cubic.InOut,

    // Quartic
    QuarticIn: _tween2.default.Easing.Quartic.In,
    QuarticOut: _tween2.default.Easing.Quartic.InOut,
    QuarticInOut: _tween2.default.Easing.Quartic.InOut,

    // Quintic
    QuinticIn: _tween2.default.Easing.Quintic.In,
    QuinticOut: _tween2.default.Easing.Quintic.InOut,
    QuinticInOut: _tween2.default.Easing.Quintic.InOut,

    // Sinusoidal
    SinusoidalIn: _tween2.default.Easing.Sinusoidal.In,
    SinusoidalOut: _tween2.default.Easing.Sinusoidal.InOut,
    SinusoidalInOut: _tween2.default.Easing.Sinusoidal.InOut,

    // Circular
    CircularIn: _tween2.default.Easing.Circular.In,
    CircularOut: _tween2.default.Easing.Circular.InOut,
    CircularInOut: _tween2.default.Easing.Circular.InOut,

    // Elastic
    ElasticIn: _tween2.default.Easing.Elastic.In,
    ElasticOut: _tween2.default.Easing.Elastic.InOut,
    ElasticInOut: _tween2.default.Easing.Elastic.InOut,

    // Back
    BackIn: _tween2.default.Easing.Back.In,
    BackOut: _tween2.default.Easing.Back.InOut,
    BackInOut: _tween2.default.Easing.Back.InOut,

    // Bounce
    BounceIn: _tween2.default.Easing.Bounce.In,
    BounceOut: _tween2.default.Easing.Bounce.InOut,
    BounceInOut: _tween2.default.Easing.Bounce.InOut
};
module.exports = exports['default'];
